bundles/org.openhab.binding.mqtt.espmilighthub/src/main/java/org/openhab/binding/mqtt/espmilighthub/internal/discovery/EspMilightHubDiscoveryService.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.openhab.binding.mqtt.espmilighthub.internal.discovery;
+
+import static org.openhab.binding.mqtt.espmilighthub.internal.EspMilightHubBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.mqtt.discovery.AbstractMQTTDiscovery;
+import org.openhab.binding.mqtt.discovery.MQTTTopicDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.io.transport.mqtt.MqttBrokerConnection;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link EspMilightHubDiscoveryService} is responsible for finding globes
+ * and setting them up for the handlers.
+ *
+ * @author Matthew Skinner - Initial contribution
+ */
+
+@Component(service = DiscoveryService.class, configurationPid = "discovery.mqttespmilighthub")
+@NonNullByDefault
+public class EspMilightHubDiscoveryService extends AbstractMQTTDiscovery {
+    protected final MQTTTopicDiscoveryService discoveryService;
+
+    @Activate
+    public EspMilightHubDiscoveryService(@Reference MQTTTopicDiscoveryService discoveryService) {
+        super(SUPPORTED_THING_TYPES, 3, true, STATES_BASE_TOPIC + "#");
+        this.discoveryService = discoveryService;
+    }
+
+    @Override
+    protected MQTTTopicDiscoveryService getDiscoveryService() {
+        return discoveryService;
+    }
+
+    @Override
+    public void receivedMessage(ThingUID connectionBridge, MqttBrokerConnection connection, String topic,
+            byte[] payload) {
+        resetTimeout();
+        if (topic.startsWith(STATES_BASE_TOPIC)) {
+            String cutTopic = topic.replace(STATES_BASE_TOPIC, "");
+            int index = cutTopic.indexOf("/");
+            if (index != -1) // -1 means "not found"
+            {
+                String remoteCode = (cutTopic.substring(0, index)); // Store the remote code for use later
+                cutTopic = topic.replace(STATES_BASE_TOPIC + remoteCode + "/", "");
+                index = cutTopic.indexOf("/");
+                if (index != -1) {
+                    String globeType = (cutTopic.substring(0, index));
+                    String remoteGroupID = (cutTopic.substring(++index, ++index));
```

Comment:
I think the use of `++index` makes this harder to understand. It doesn't appear that incrementing the index even matters here, so please just rewrite this without using `++`.

Reply:
It is needed. I could re-write it as this which does the same thing and may be easier to understand.

String remoteGroupID = (cutTopic.substring(index+1, index+2));

index++ This means use the variable and then increment it AFTER the value is used.
++index This means increment the variable BEFORE it is used.

Reply:
I'm aware of the difference between `index++` and `++index`. The issue I have is that you are using `++index` for each parameter. Knowing the evaluation order ( not only the `++index` but of each of the parameters relative to each other ) is more knowledge than I think is necessary for an eventual code maintainer to be required to have.
```suggestion
                    String remoteGroupID = (cutTopic.substring(index+1, index+2));
```
Keep it simple please.